enhancements/network/ovn-hypershift-live-migration.md

```diff
@@ -0,0 +1,303 @@
+The CNI will not set an address at virt-launcher pod netns, that address is
+assigned to the VM with the DHCP options from the LSP, this allows to use 
+kubevirt bridge binding with pod networking and still do live migration.
+
+#### IPAM
+
+The point to point routing feature allows an address to be running at a node different from
+the one "owning" the subnet the address is coming from. This will happen
+after VM live migration.
+
+One scenario for live migration is to shut down the node the VMs were migrated
+from, this means that the IPAM node subnet should go back to the pool but since
+the migrated VMs contains IPs from it those IPs should reserved in case the 
+subnet is assigned to a new node.
+
+On that case before assigning the subnet to the node the VMs ips need to be 
+reserved so they don't get assigned to new pods
+
+Another scenario is ovn-kubernetes pods restarting after live migration, on 
+that case ovnkube-master should discover to what ip pool the VM belongs 
+(using an annotation or searching for subnets) and reserve the address.
+
+#### Detecting hypershift VMs and live migration
+
+This enhancement only applies for hypershift kubevirt workers so the rest of 
+virtual machine should not use it, those VMs will be labeled with
+`kubevirt.io/allow-pod-network-bridge-binding-live-migration` by hypershift
+kubevirt provider and ovn-k will do point to point routing if detected, the of non
+hypershift VM should not add that special label.
+
+The VM's host node change after live migration and the traffic has to be 
+redirected over this host node. To do so the point to point routing has to be
+changed and to detect post live-migration event a pair of 
+kubevirt virt-launcher pod labels has to be inspected, `kubevirt.io/nodeName` 
+and `creationTimestamp`, the first one is set after live migration with the 
+name of the new host node and the second one allow ovn-k to disregard old
+virt-launcher pod (they are not deleted after live-migration).
+
+### Risks and Mitigations
+
+#### OVN extended arp_proxy features
+
+
+The current ovn arp_proxy is missing the following features: enforcing the mac 
+address and ipv6, ideally those get implemented and ovn-kubernetes have 
+them for 4.14 release.
+
+In case they don't arrive on time, it will not be able to configure IPv6 hosted
+clusters at hypershift and VMs arp table will need to be refreshed after live
+migration. 
+
+### Drawbacks
+
+Doing live migration over pod network with kubevirt and changing its topology may present some risks and need to be 
+thoroughly tested and supported.
+
+### Open Questions 
+
+Do this impact non hypershift payloads ? The control-plane will have some more
+load from the point to point routing but data-plane is not affected
+
+Do this work with future interconnect topology ? It's not using a distributed 
+switch per hosted cluster so that part should be fine, but it may needed 
+to replicate point to point routing at all node's ovn_cluster_router.
+
```

Comment:
Additional question. What kind of scale do you expect here? If we had 1000s of these migrated pods we could have 1000s of new policies and routes. With IC that would be 1000s*nodes.

Reply:
The scale is on thousands to reduce control plane preasure (with IC there will be one ovn control plane per node) an addres set per node can be used, it will group all the IPs belonging to kubevirt VM's for a specific node so we will have one policy/route per node.

This will drastically reduce the number of logical flows but the number of ovs flows will be the same, after talking with some OVN maintainers looks like for the ovs part is not that bad.

Is the address set worth it ? Do we need to do it at this enhancement or can it be done at follow up ?

Reply:
>Also anoter optimization is that the point to point routing is only needed if the VM's is running at a node that do not "own" the VM's IP, meaning that if there is no live-migration we don't need point to point routing for the VM, this make sense ?

I dont think we should have static routes for pods that are not live migrated. The regular routes should continue to function like they do today.

Reply:
I will state that at the doc, generally speaking point to point routing has to be specified if the pod IP dos not belongs to the node's subnet, since it can happend that we do a second live migration and the VM goes back to the original node so in that case the policies and static routes are not needed.

Reply:
> I dont think we should have static routes for pods that are not live migrated. The regular routes should continue to function like they do today.

@trozet @maiqueb I want to caution against this as I think the complexity of having two modes (normal routing when on original node vs /32 after migration) might outweigh the value of this optimization. I'm not entirely sure this optimization will do anything in our use case either.

Here's the reason. While it's true that when guest clusters hosted on KubeVirt VMs are initially created, the VMs will all be on the original node they started on, possibly for quite some time. The issue here is that the moment someone updates the underlying infra cluster, this will nearly 100% change. Every single VM within the cluster will Live Migrate as the infra cluster nodes are rebooted during the update. At that point all the pods will shuffle around, and this optimization is nearly completely void.

Reply:
while that's true, I think all we're saying is the reconcile cycles should dispose of that route + policy when they're not needed.

@davidvossel